Unity-Technologies/Graphics
# Added shadowmask and shadowmask distance support

### Checklist for PR maker

---
### Purpose of this PR
- Added shadowmask and shadowmask distance support.
- Renaming _MIXED_LIGHTING_SUBTRACTIVE into LIGHTMAP_SHADOW_MIXING. This gives us few advantages: consistency with builtin/hdrp, re-using existing builtin keyword, keyword stripping is automatically handled by core, we use keyword of core functionality.
- Refactoring lightOcclusionChannel to work as lightShadowMaskSelector (Where we used mask). Sadly I can not renamed property as it would break public API.
- Refactoring unity_ProbesOcclusion to be applied in MixRealtimeAndBakedShadows instead of light.distanceAttenuation. This way keeping similar functionality to builtin. Now light probes behaves more closely what we have in builtin.
- Changing shadow fade to be calcualte inside MixRealtimeAndBakedShadows instead of in Light gathering. As this way required by realtime and baked shadow mixing.

# **For more information please check the card https://jira.unity3d.com/browse/URP-36!**

---
### Testing status

**Manual Tests**: What did you do?
- Other: 

**Comparing URP and Builtin mixed lighting. Simple scene with static walls, dynamic cube and lightprobes group.**:

*Builtin Mixed Lighting Subtractive + Probes*
- Static mesh keywords LIGHTMAP_ON LIGHTMAP_SHADOW_MIXING.
- Static mesh effected by LIGHMAP + SCREENSPACE_SHADOWS.
- Dynamic mesh keywords LIGHTMAP_SHADOW_MIXING.
- Dynamic mesh effected by PROBES + SCREENSPACE_SHADOWS.
![image](https://user-images.githubusercontent.com/24258017/91164715-371a0d00-e6d8-11ea-962a-cb71110efcf7.png)
*URP Mixed Lighting Subtractive + Probes*
![image](https://user-images.githubusercontent.com/24258017/91165816-0509aa80-e6da-11ea-9c6b-e645d9b84a7f.png)

*Builtin Mixed Lighting ShadowMask + Probes*
- Static mesh keywords LIGHTMAP_ON LIGHTMAP_SHADOW_MIXING SHADOWS_SHADOWMASK.
- Static mesh effected by SHADOWMASK + SCREENSPACE_SHADOWS.
- Dynamic mesh keywords LIGHTMAP_SHADOW_MIXING.
- Dynamic mesh effected by PROBES + SCREENSPACE_SHADOWS.
![image](https://user-images.githubusercontent.com/24258017/91166527-2cad4280-e6db-11ea-95c4-ea2e6ac5d286.png)
*URP Mixed Lighting ShadowMask + Probes*
![image](https://user-images.githubusercontent.com/24258017/91166742-801f9080-e6db-11ea-88d1-f02a3da05eba.png)

*Builtin Mixed Lighting ShadowMask + Distance + Probes*
- Static mesh keywords LIGHTMAP_ON SHADOWS_SHADOWMASK.
- Static mesh effected by SHADOWMASK + SCREENSPACE_SHADOWS.
- Dynamic mesh keywords None.
- Dynamic mesh effected by PROBES + SCREENSPACE_SHADOWS. (Probes turn only when realtime is out of distance)
![image](https://user-images.githubusercontent.com/24258017/91167508-bf9aac80-e6dc-11ea-9567-cf9589d6343f.png)
*URP Mixed Lighting ShadowMask + Probes* (Has harder cut-off, but it is not part of shadowmask change. However it is something we should look into.)
![image](https://user-images.githubusercontent.com/24258017/91167543-cc1f0500-e6dc-11ea-8855-1c7ae19f459b.png)

**Automated Tests**:
049_Lighting_Mixed_Subtractive
![image](https://user-images.githubusercontent.com/24258017/91161356-55c9d500-e6d3-11ea-8e34-8da8a8680dd5.png)

043_Lighting_Mixed_ShadowMask
![image](https://user-images.githubusercontent.com/24258017/91161475-77c35780-e6d3-11ea-97fd-1bcd0b218714.png)

134_Lighting_Mixed_ShadowMask_Distance
![image](https://user-images.githubusercontent.com/24258017/91161576-975a8000-e6d3-11ea-9280-aae9ae79ce8a.png)

**Yamato**:
https://yamato.prd.cds.internal.unity3d.com/jobs/902-Graphics/tree/universal%252Fshadow-mask-support

Any test projects to go with this to help reviewers?

---
### For QA
- Check if mixed lighting subtractive did not regressed in any way.
- Check if mised lighting shadowmask and shadowmask distance works similar to builtin.
- Make sure keywords are stripped if mixed lighting is not used (Either disabled in pipeline asset or not enabled in lighting settings or not baked).
